odoo/addons/base/models/res_partner.py

```diff
@@ -374,6 +374,8 @@ def onchange_email(self):
     def _onchange_company_id(self):
         if self.parent_id:
             self.company_id = self.parent_id.company_id.id
+        if self.company_id and self.company_id not in self.user_id.company_ids:
```

Comment:
Maybe ``if self.user_id and self.company_id and self.company_id not in self.user_id.company_ids`` ?